Schema for database

ttung opened the pull request:
### Description

#### Issue
[ch68091](https://app.clubhouse.io/genepi/stories/space/68091)

### Test plan
it's a doc. :)

ttung commented:
You can paste the text into https://dbdiagram.io/d to visualize it.

ttung commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,115 @@
+///////////////////////////////////////////////////
+// DPHs and users
+
+Table Dph {
+  id int [pk, increment]
+  name varchar [not null, unique]
+  gisaid_division_name varchar [unique]
+}
+
+Table User {
+  id int [pk, increment]
+  name varchar [not null]
+  auth0_user_id varchar [not null]
+  dph int [not null, ref: > Dph.id]
+  dph_admin bool [not null]
+}
+
+// members of dph can see data uploaded by owner_dph
+Table DphCanSee {
+  dph int [not null, unique, ref: > Dph.id]
+  owner_dph int [not null, unique, ref: > Dph.id]
+}
```
@sidneymbell  could you comment on this?

phoenixAja replied:
what is the difference between owner_dph and dph_admin?

ttung replied:
I should rename that as:

viewing_dph
owner_dph

Meaning viewing_dph can see the data for owner_dph.

phoenixAja replied:
I'm also not sure if we would need the `DphCanSee` table since we could determine whether a user was in a dph or an admin by looking at the `User` table directly.

ttung replied:
`DphCanSee` is for people to cross that boundary.

phoenixAja replied:
sorry i'm not really sure what you mean by that. Why not just query the User and see what DPH they belong to (and if they are an admin) directly in the User table? If they are a part of the DPH then they can see the data?

ttung replied:
So for instance, we might have an entry for people in CADPH to view SCCDPH's data.

phoenixAja replied:
ok got it thanks

ttung replied:
I'm not fully sold this is the right model, because it doesn't fully capture the nuances.  For instance, SCCDPH might want to let CADPH view its data but not its metadata.  But making it a full cross product might be a bit too heavyweight.

ttung commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,115 @@
+  }
+}
+
+Table Gene {
+  id int [pk, increment]
+  name varchar [not null, unique]
+}
+
+Table PcrPrimer {
+  id int [pk, increment]
+  gene_id int [not null, ref: > Gene.id]
+  name varchar [not null]
+  
+  Indexes {
+    (gene_id, name) [unique]
+  }
+}
+
+Table PcrCtValue {
+  physical_sample_id int [not null, ref: > PhysicalSample.id]
+  pcr_primer_id int [not null, ref: > PcrPrimer.id]
+  value float [not null]
+  
+  Indexes {
+    (physical_sample_id, pcr_primer_id) [unique]
+  }
+}
```
@jackkamm would love your feedback on if this is adequate to capture CT values but is flexible enough for changes in the PCR assay.

jackkamm replied:
Honestly I'm a little hesitant to store all these CT values. There are so many assays, they are so heterogeneous, some of them don't even give CT values but give other readouts. While they have been generally useful for us in sanity checking our sequencing runs, for covidtracker v2 I think we will devolve this responsibility to the submitting DPHs

ttung replied:
@sidneymbell if you're on board with this, i can drop this.

jackkamm commented:
Random thoughts/questions:

Is `RawSequence` the fastqs?

Should store info about what sequencer was used. In particular, it will be important to distinguish illumina vs oxford nanopore data. Also, what protocol was used -- typically this will be ARTICv3 but this could always change in future

I recommend storing the BAM file intermediates between the fastqs and the fastas as well

I recommend storing some sequence quality statistics -- main ones are depth of coverage, number of sites recovered, number of Ns, and number of mixed/ambiguous sites

I'll be out tomorrow but can give more feedback later in week on Thursday or Friday

jackkamm commented:
Another comment: would be good to store NCBI names and accession IDs for each of the uploaded fastqs, bams, and fastas.

Speaking of which -- we'd need to upload host-filtered fastqs to SRA, so perhaps there should be an additional object for the host-filtered fastqs.

ttung commented:
> Is RawSequence the fastqs?

yes.

> Should store info about what sequencer was used. In particular, it will be important to distinguish illumina vs oxford nanopore data. Also, what protocol was used -- typically this will be ARTICv3 but this could always change in future

what fields do we need?  should sequencer be free text or constrained?  protocol?

> I recommend storing the BAM file intermediates between the fastqs and the fastas as well

Do the BAM files have human read data?

> I recommend storing some sequence quality statistics -- main ones are depth of coverage, number of sites recovered, number of Ns, and number of mixed/ambiguous sites

these are all ints, and per Sequence, right?  Do we expect users to upload this?

lvreynoso commented:
Perhaps we could filter out human reads before doing any analysis? And we could treat the filtered fastq as the "input" file, and only store that. Then human read data should be excluded from any other output files, right?

lvreynoso commented:
And the sequence quality stats would be generated by us, not the users. Depth and coverage are usually decimals/fractional.

ttung commented:
> And the sequence quality stats would be generated by us, not the users. Depth and coverage are usually decimals/fractional.

Wouldn't that be generated by users if they're uploading fastas?

lvreynoso commented:
Fastas/fastqs come with *read quality* information for individual reads, but not higher order quality analysis. Those statistics are only obtainable by doing further data analysis (e.g. running STAR, PriceSeq, etc.) which presumably DPHs wouldn't be doing since they would just be sequencing samples and sending us the fastqs.

sidneymbell commented:
Haven't had a chance to look (OOO today) but will in the morning

On Wed, Jan 27, 2021 at 5:36 PM Tony Tung <notifications@github.com> wrote:

> *@ttung* commented on this pull request.
> ------------------------------
>
> In docs/backend/schema.ddl
> <https://github.com/chanzuckerberg/covidr/pull/30#discussion_r565760601>:
>
> > +Table Gene {
> +  id int [pk, increment]
> +  name varchar [not null, unique]
> +}
> +
> +Table PcrPrimer {
> +  id int [pk, increment]
> +  gene_id int [not null, ref: > Gene.id]
> +  name varchar [not null]
> +
> +  Indexes {
> +    (gene_id, name) [unique]
> +  }
> +}
> +
> +Table PcrCtValue {
> +  physical_sample_id int [not null, ref: > PhysicalSample.id]
> +  pcr_primer_id int [not null, ref: > PcrPrimer.id]
> +  value float [not null]
> +
> +  Indexes {
> +    (physical_sample_id, pcr_primer_id) [unique]
> +  }
> +}
>
> @sidneymbell <https://github.com/sidneymbell> if you're on board with
> this, i can drop this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chanzuckerberg/covidr/pull/30#discussion_r565760601>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADAIYXYNUVYK6KL25QT7W5TS4C5QFANCNFSM4WUILA7A>
> .
>

jackkamm commented:
> Wouldn't that be generated by users if they're uploading fastas?

Are we allowing people to upload their own fastas if they ran their own bioinformatics? Or are we only allowing them to upload fastqs of reads?

In the former case, we would want them to upload some of these quality statistics; otherwise we will generate ourselves when running the bioinformatics pipelines.

phoenixAja commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,115 @@
+///////////////////////////////////////////////////
+// Phylo Runs
+
+Table PhyloRun {
+  id int [pk, increment]
+}
```
should we also have a field here that points to the location of the output JSONs?

ttung replied:
LOL right.  :)

phoenixAja commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,115 @@
+  s3_path varchar [not null, unique]
+}
+
+Table Sequence {
+  id int [pk, increment]
+  gisaid_name varchar [unique]
+  gisaid_accession varchar [unique]
+
+  // TODO: Do we store the genome directly on this object?
+}
+
+// describes a single sequence that was uploaded to the system directly.
+Table UploadedSequence {
+  sequence_id int [not null, unique, ref: > Sequence.id]
+  physical_sample_id int [not null, ref: > PhysicalSample.id]
+}
```
why do we need separate tables for `Sequence` and `UploadedSequence`?

ttung replied:
`Sequence` is like the parent object for `UploadedSequence` and `ConsensusGenomeSequence`.  There needs to be two ways for a sequence to be connected to a physical sample.  If the user is uploading a fasta, then the metadata is immediately connected to the fasta.  If the user is uploading a fastq, then the metadata is connected to the fastq.  After the consensus genome is called, then the consensus genome is connected to the fastq (plausibly more than once if the pipeline is called multiple times).

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+///////////////////////////////////////////////////
+// DPHs and users
```
Not a blocker, but eventually want to generalize from "DPH" --> "Group" or something like this.

ttung replied:
Done.

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+  date_taken datetime [not null]
+  external_accession varchar [not null]
+  location varchar [not null]
+  extraction_method varchar [not null]
```
This seems LIMS-y to me. @shanaxel42?

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+  Indexes {
+    (gene_id, name) [unique]
+  }
+}
```
I think of PCR primer as an attribute of `protocol`, relevant only for consensus genome generation

ttung replied:
Dropped.

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+Table PcrCtValue {
+  physical_sample_id int [not null, ref: > PhysicalSample.id]
+  pcr_primer_id int [not null, ref: > PcrPrimer.id]
+  value float [not null]
```
This also seems LIMS-y to me.

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+
+  // TODO: @jackkamm suggested storing the sequencer information.  "In particular, it will be important to distinguish
+  // illumina vs oxford nanopore data. Also, what protocol was used -- typically this will be ARTICv3 but this could
+  // always change in future"
```
Yes (see note on `primers` above)

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+  id int [pk, increment]
+  gisaid_name varchar [unique]
+  gisaid_accession varchar [unique]
+  sequence varchar [not null]
```
plus`external_accession`?

As a comp bio, when working with this data, I'd want the following information easily available as a "record":
* Public ID (currently called `gisaid_name`)
* Private ID (currently called (`external_accession`)
* Consensus genome
* Pointer to metadata (location + collection date)

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+// describes a single pathogen sequence
+Table PathogenSequence {
+  id int [pk, increment]
+  gisaid_name varchar [unique]
```
`gisaid_name` -> `public_id`

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+// describes a single sequence that was uploaded to the system directly.
+Table UploadedPathogenSequence {
+  pathogen_sequence_id int [not null, unique, ref: > PathogenSequence.id]
+  physical_sample_id int [not null, ref: > PhysicalSample.id]
```
The vast majority of users won't have this

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+  pathogen_sequence_id int [not null, unique, ref: > PathogenSequence.id]
+  raw_genome_id int [not null, ref: > RawSequence.id]
+
+  // TODO: @jackkamm suggested storing the intermediate files (BAM file?, host-filtered short reads).  Need more
```
I'd put this in the same bucket as storing the fastq files -- good to keep on hand in case something needs to be reprocessed, but we don't need to store indefinitely

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+  // information how to work that into the schema.
+
+  // TODO: @jackkamm suggested storing quality statistics.  "main ones are depth of coverage, number of sites recovered,
+  // number of Ns, and number of mixed/ambiguous sites"
```
Yes.

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+
+  // metadata about the phylo run
+  run_date datetime [not null]
+  // TODO: additional metadata?  sw version?  information about the gisaid data dump.
```
Yeah. I'd potentially even suggest having a more generic `analysis` table that just lists out
1. type of analysis (consensus or phylo)
2. pipeline and version
3. date of run
4. any errors that were raised
5. pointers for where to find the output files (including raw results files like `tree.newick` and `aligned.fasta`)

sidneymbell replied:
Idea being that eventually the api for a generic pipeline module looks something or other like:

input: 
`records` as described above
` parameters` (out of scope for now, but eventually)
`output` (where output files should be stored

output:
the above ^ 
`intermediate results` --> `tree.newick`, etc. to be stored
`end product` --> `build.json`, `consensus genome`, etc.

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,142 @@
+  }
+}
+
+// this enables a many:many relationship between Sequence and PhyloRun.
```
Yussssss

jackkamm commented:
I suggest referring to these for designing the schema:
- https://github.com/pha4ge/SARS-CoV-2-Contextual-Data-Specification/blob/master/PHA4GE%20SARS-CoV-2%20Contextual%20Data%20Template.xlsx
- https://github.com/seattleflu/documentation/blob/master/id3c-warehouse-schema.pdf

jackkamm commented:
Another good reference is COG-UK's system

https://www.biorxiv.org/content/10.1101/2020.10.06.328328v1

![image](https://user-images.githubusercontent.com/9089216/106333433-7c210d00-623d-11eb-928a-8621b720b313.png)

ttung commented:
I made a bunch of changes to this.  Please review when you have time. :)

ttung commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,280 @@
+///////////////////////////////////////////////////
+// Consensus genome pipeline
+
+// TODO: @jackkamm do the BAM/SRA records need to store any metrics?
```
@jackkamm

jackkamm replied:
These are recommended by PHA4GE, and nice to have, but I think not strictly necessary:
- library insert size (int)
- number base pairs (int)
- mean contig length (float)
- N50 (int)

ttung replied:
Added, thanks!!

erinstargazer commented:
I am just getting up to speed on this so don't have many comments since I'm mainly still learning it.

I was wondering about the situation with `Entity` though. It looks like it's being used as an abstraction for specific relationships (I might just be way off the mark here though). For instance it seems like `ProcessedGisaidDump` would have an associated `RawGisaidDump` but the only way I could see that happening is through the `Entity` relationships they both have. At some point `Entity` becomes so generalized and it feels like almost everything could be an `Entity`. This can lead to less information being conveyed. For instance, if I have a car with 4 tires, having a relationship that each tire can go to 1 car is clearer than having each tire having an `Entity` and each car also having an `Entity` and then needing to infer the rest of the info. Sorry if this is nonsense but I wanted to find out more of the motivation and guiding principles around the `Entity` abstraction.

ttung commented:
So entity being the base unit is motivated by two things:
1. A lot of things may be submitted to public repositories.
2. It's a way to glue workflows from beginning to end.

The relationship between entities is a bit more complicated mostly because of the nature of workflows.  You'd have to find the workflows and find the corresponding outputs.  There isn't a table that directly maps input to output.

I think the solution is to provide helper methods in each of the specific entities that simplify finding the relationships.  For instance, `RawGisaidDump` might have a `processed_dumps()` method that abstracts that.

jackkamm commented:
Getting the following error on most recent commit in dbdiagram.io

```
 (64:5) Column "submitting_group" do not exist in table "PhysicalSample" 
```

![image](https://user-images.githubusercontent.com/9089216/106525291-9c98d380-6498-11eb-91b1-c21dfba9be26.png)

jackkamm commented:
It's fixed by the following patch
```
   // TODO: the other fields from PHA4GE
 
   Indexes {
-    (submitting_group, private_identifier) [unique]
+    (submitting_group_id, private_identifier) [unique]
   }
 }
```

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+  website VARCHAR [unique]
+}
+
+Table Accession {
```
I think this table needs a column for the actual accession (which should be a varchar)

ttung replied:
HAHA yeah you're totally right.

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+}
+
+// describes a single set of sequence data from the sequencer. (FASTQ)
+Table RawSequence {
```
Is this the raw fastq or the host-filtered fastq?

If the former, we need another object for the host-filtered fastq.

If we host-filter the fastq upfront then we could just store that, and not store the raw fastq. (Similar to how idseq does it)

Also, this needs to be linked to the BAM

ttung replied:
I did not realize there were two. :)

ttung replied:
IDseq doesn't store the raw upload?  Did not know that.

jackkamm replied:
Not 100% sure about how IDseq stores the fastqs, I might be wrong about it

sidneymbell replied:
we should 100% be filtering out host reads and discarding them; these should never be stored

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+  sequencing_date FLOAT
+
+  // optional field for gisaid submission, we'd like to get users to provide this.
+  sequencing_depth FLOAT
```
This should be a column of the BAM not the fastq

ttung replied:
I thought this was user-provided data.  In that case, it would need to be attached to the fastq object.

jackkamm replied:
No it's the average number of reads covering each position in the reference and is computed by aligning reads to reference. The user needs to supply it if they upload a fasta, but if they upload a fastq we compute it ourselves

sidneymbell replied:
While this _can_ be user provided if they're uploading a fasta, the vast majority will not have this information. 

If they upload a fastq, we'll compute it ourselves.

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+Table Group {
+  id INT [pk, increment]
+  name VARCHAR [not null, unique]
+  gisaid_division_name VARCHAR [unique]
```
I don't think it is a safe assumption that each Group will be 1-1 with a location. For example, there is a single public lab for Napa, Sonoma, Yolo, Marin, which ideally should be separate locations in GISAID.

Also, I think this might be redundant with the PhysicalSample.location column for location of physical samples.

On the other hand, we do need to provide an address for the group, for submission to GISAID

ttung replied:
Will fix.

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+
+  // TODO: dedup these fields with PHA4GE
+  date_taken DATETIME [not null]
+  location VARCHAR [not null]
```
Is this always going to be a county?

Also, will we ever include samples from other states/countries? If so we should include columns for these.

Perhaps instead of a single VARCHAR, we could link to a separate table that would store county, state, country and perhaps other info (lat/lon?)

ttung replied:
Sounds good.

sidneymbell replied:
IDSeq has a very nice auto-suggest feature in their upload process where the user starts typing a location (address, city, county), and it provides discrete (read: standardized) options and "rounds" these values to the appropriate level of detail for confidentiality (i.e., county or less resolution)

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+// describes a sequencer machine  (TODO: what's a good way of canonicalizing this?)
+Table Sequencer {
+  id INT [pk, increment]
+  name VARCHAR [not null, unique]
```
Here are the valid Sequencer names that NCBI will accept. I am not sure where/how we should encode them specifically

- Illumina HiSeq 1000
- Illumina HiSeq 1500
- Illumina HiSeq 2000
- Illumina HiSeq 2500
- Illumina HiSeq 3000
- Illumina HiSeq 4000
- Illumina iSeq 100
- Illumina NovaSeq 6000
- Illumina MiniSeq
- Illumina MiSeq
- NextSeq 500
- NextSeq 550

ttung replied:
What about nanopore sequencing?

jackkamm replied:
You're right sorry, here are the valid values (taken from the PH4AGE spreadsheet I attached above)

```
sequencing instrument
Illumina
    HiSeq X
    HiSeq X Five
    HiSeq X Ten
    Illumina Genome Analyzer
    Illumina Genome Analyzer II
    Illumina Genome Analyzer IIx
    Illumina HiScanSQ
    Illumina HiSeq 1000
    Illumina HiSeq 1500
    Illumina HiSeq 2000
    Illumina HiSeq 2500
    Illumina HiSeq 3000
    Illumina HiSeq 4000
    Illumina iSeq 100
    Illumina NovaSeq 6000
    Illumina MiniSeq
    Illumina MiSeq
    NextSeq 500
    NextSeq 550
Pacific Biosciences
    PacBio RS
    PacBio RS II
    PacBio Sequel
    PacBio Sequel II
Ion Torrent
    Ion Torrent PGM
    Ion Torrent Proton
    Ion Torrent S5 XL
    Ion Torrent S5
Oxford Nanopore Technologies
    GridION
    MinION
    PromethION
BGI Genomics
    BGISEQ-500
MGI
    DNBSEQ-T7
    DNBSEQ-G400
    DNBSEQ-G400 FAST
    DNBSEQ-G50
Not Applicable
Not Collected
Not Provided
Missing
Restricted Access
```

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+  physical_sample_id INT [not null, ref: > PhysicalSample.id]
+
+  // optional field for gisaid submission, we'd like to get users to provide this.
+  sequencing_depth FLOAT
```
I think this should be mandatory not optional

ttung replied:
I thought we looked it up and it was an optional field for gisaid.

jackkamm replied:
We should test uploading a sequence to GISAID without it, but it's the most basic sequence quality statistic so I think we should require it

jackkamm replied:
Or if not require, then strongly strongly encourage -- perhaps by making it mandatory to start, and then only consider loosening this if a submitter can make a compelling case why this would be too onerous for them to collect and submit

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+
+// TODO: @jackkamm do the BAM/SRA records need to store any metrics?
+
+Table HostFilteredBam {
```
`HostFiltered` is redundant since it's aligned to SARS-CoV-2. Just call it `Bam` or `AlignedReads`.

Also, this needs to be linked to the `RawSequence` and the `CalledPathogenSequence` somehow

jackkamm replied:
Add a column for the reference genome accession

ttung replied:
> `HostFiltered` is redundant since it's aligned to SARS-CoV-2. Just call it `Bam` or `AlignedReads`.

👍 

> Also, this needs to be linked to the `RawSequence` and the `CalledPathogenSequence` somehow

It's done through the workflow objects.

> Add a column for the reference genome accession

Anything that is an entity can have an accession (which you noted was missing).

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+
+// describes a single sequence that was called from a consensus genome run
+Table CalledPathogenSequence {
+  pathogen_sequence_id INT [not null, unique, ref: > PathogenSequence.id]
```
This needs to be linked to the Bam

jackkamm replied:
Add column for the software used to generate the sequence. Needed for submission to GISAID/NCBI

PH4AGE actually recommends separating this out into separate fields for the software used for raw sequence data processing, assembly method, and consensus sequence method. I think the first 2 may only be relevant for long-read (Oxford Nanopore) data but I'm not sure -- should check in with someone who knows more about the long-read genome pipelines.

ttung replied:
> This needs to be linked to the Bam

It is.  This should is the output of a workflow, so you can trace it back from the Workflow to the BAM.

> Add column for the software used to generate the sequence. Needed for submission to GISAID/NCBI

> PH4AGE actually recommends separating this out into separate fields for the software used for raw sequence data processing, assembly method, and consensus sequence method. I think the first 2 may only be relevant for long-read (Oxford Nanopore) data but I'm not sure -- should check in with someone who knows more about the long-read genome pipelines.

The workflow object contains the software information.

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+///////////////////////////////////////////////////
+// Physical Samples
+
+Table PhysicalSample {
```
This needs a column for the `isolate` (the name used for submission to GISAID/Genbank/etc)

jackkamm replied:
We should also add a column for `purpose of sampling` as recommended by PH4AGE. Important to distinguish between samples collected for surveillance, for outbreak investigation, and for targeted searches e.g. S-gene dropout

ttung replied:
`isolate` would all be SARS-CoV-2, right?  this is futureproofing, I assume?

jackkamm replied:
No, `isolate` is what we've been calling `gisaid_name` -- basically, it's the human-readable public name of the sample, and is usually something like `USA/CA-CZB-1234/2021`

ttung commented:
> It's fixed by the following patch

thanks! patched.

jackkamm commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,281 @@
+}
+
+// describes a single sequence that was uploaded to the system directly.
+Table UploadedPathogenSequence {
```
Add column for the software used to generate the sequence. Needed for submission to GISAID/NCBI

PH4AGE actually recommends separating this out into separate fields for the software used for raw sequence data processing, assembly method, and consensus sequence method. I think the first 2 may only be relevant for long-read (Oxford Nanopore) data but I'm not sure -- should check in with someone who knows more about the long-read genome pipelines.

jackkamm replied:
Add a column for the reference genome

ttung replied:
> Add column for the software used to generate the sequence. Needed for submission to GISAID/NCBI

> PH4AGE actually recommends separating this out into separate fields for the software used for raw sequence data processing, assembly method, and consensus sequence method. I think the first 2 may only be relevant for long-read (Oxford Nanopore) data but I'm not sure -- should check in with someone who knows more about the long-read genome pipelines.

are these mandatory fields?

> Add a column for the reference genome

is this mandatory? Also, is this a choice of a few acceptable reference genomes?  or is it freeform text?

jackkamm replied:
I think consensus sequence method is mandatory. In any case it's really important information to provide publicly. For us, a minimal version of it is `ivar=1.2`

jackkamm reviewed (changes requested):
I've cross-checked with PHA4GE. Please see the attached spreadsheet. I've also made inline comments in the relevant locations.

[PHA4GE-aspen-crosscheck.xlsx](https://github.com/chanzuckerberg/aspen/files/5907741/PHA4GE-aspen-crosscheck.xlsx)

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+
+Table Group {
+  id INT [pk, increment]
+  name VARCHAR [not null, unique]
+  email VARCHAR [not null, unique]
+  address VARCHAR [not null, unique]
```
Referring to group owner?

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+Table TypeOfData {
+  // Types of data people can view.  This table should be pre-populated with at least the following entries:
+  //  1. Trees
+  //  2. Sequences
```
3. metadata (per column)

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+
+  // This is the public identifier we assign to this sample.  The identifier assigned by the public repositories is
+  // stored on the related Accession objects.
+  public_identifer VARCHAR [not null, unique]              // maps to specimen_collector_sample_id and isolate
```
Re: comment on `isolate` above... are these the same things? (I personally find it more intuitive to think of this as `public_identifier`, although agree we should take the ph4ge terminology for the sake of being good citizens

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+  collector VARCHAR [not null]                             // maps to sample_collected_by
+  collector_email VARCHAR                                  // maps to sample_collector_contact_email
+  collector_address VARCHAR                                // maps to sample_collector_contact_address
+  sequencer VARCHAR                                        // maps to sequence_submitted_by
```
I first read this as what instrument model was used

ttung replied:
Commented out. :)

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+  sequencer_address VARCHAR                                // maps to sequence_submitter_contact_address
+
+  collection_date DATETIME [not null]                      // maps to sample_collection_date
+  received_date DATETIME                                   // maps to sample_received_date
```
If we're not doing the sequencing anymore, we no longer have a `received_date` for physical samples?

ttung replied:
Dropped.

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+  collection_date DATETIME [not null]                      // maps to sample_collection_date
+  received_date DATETIME                                   // maps to sample_received_date
+  location VARCHAR [not null]
+  state VARCHAR [not null]                                 // maps to geo_loc_name_state_province_region
```
```suggestion
  division VARCHAR [not null]                                 // maps to geo_loc_name_state_province_region
```

sidneymbell replied:
this is what nextstrain uses as a catchall for `state` `province` etc

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+  country VARCHAR [not null]                               // maps to geo_loc_name_country
+  country_residence VARCHAR                                // maps to host_origin_geo_loc_country
+  country_exposure VARCHAR                                 // maps to location_of_exposure_geo_loc_name_country
+  travel_history VARCHAR                                   // maps to travel_history
```
potential pii

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+  host_sex ENUM                                            // maps to host_gender
+  symptom_onset_date DATETIME                              // maps to symptom_onset_date
+  signs_and_symptoms VARCHAR                               // maps to signs_and_symptoms
+  exposure_event VARCHAR                                   // maps to exposure_event
```
potential phi

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+///////////////////////////////////////////////////
+// Physical Samples
+
+Enum Sex {
+  Female
+  Male
+  Unknown
+}
```
potential pii

ttung replied:
Removed.

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+Table WorkflowInputs {
+  entity_id INT [not null, ref: > Entity.id]
+  workflow_id INT [not null, ref: > Workflow.id]
+
```
are we logging the many-to-many relationship of samples <-> workflow_runs?

ttung replied:
Yes.

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+  name VARCHAR [not null, unique]
+}
+
+Table Workflow {
```
`workflow_run`?

ttung replied:
Clarified in the meeting.

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+Table WorkflowIntermediateEntities {
+  entity_id INT [not null, ref: > Entity.id]
+  workflow_id INT [not null, ref: > Workflow.id]
+
```
does this then get intermediate output files attached to it?

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+  }
+}
+
+Table WorkflowIntermediateEntities {
```
```suggestion
Table WorkflowIntermediateOutputs {
```

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+///////////////////////////////////////////////////
+// Entities and Workflows
+
+Table Entity {
```
```suggestion
Table Workflow {
```

ttung replied:
Clarified in the meeting.

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+  website VARCHAR [unique]
+}
+
+Table Accession {
```
how does this link back to the `private_id`, etc?

sidneymbell commented on docs/backend/schema.ddl:
```diff
@@ -0,0 +1,342 @@
+// Sequences
+
+// describes a sequencer machine  (TODO: what's a good way of canonicalizing this?)
+Table Sequencer {
```
`Instrument`...? Idk

Commit:
Schema for database

### Description

#### Issue
[ch68091](https://app.clubhouse.io/genepi/stories/space/68091)

### Test plan
it's a doc. :)
